controlplane/kubeadm/controllers/helpers.go

```diff
@@ -246,3 +248,87 @@ func (r *KubeadmControlPlaneReconciler) generateMachine(ctx context.Context, kcp
 	}
 	return nil
 }
+
+// MachinesNeedingRollout return a list of machines that need to be upgraded.
+func (r *KubeadmControlPlaneReconciler) MachinesNeedingRollout(ctx context.Context, c *internal.ControlPlane) internal.FilterableMachineCollection {
+	machinesToUpgrade := internal.NewFilterableMachineCollection()
+
+	now := metav1.Now()
+	if c.KCP.Spec.UpgradeAfter != nil && c.KCP.Spec.UpgradeAfter.Before(&now) {
```

Comment:
Let's add comments on each if condition we have here and down when we loop through the machines to explain why.